Flaky test in crypto package: TestDSA

Every so often the `TestDSA` test fails

```
?	github.com/hkparker/go-i2p	[no test files]
ok      github.com/hkparker/go-i2p/lib/common   0.007s  coverage: 77.8% of statements
?       github.com/hkparker/go-i2p/lib/common/base32    [no test files]
?       github.com/hkparker/go-i2p/lib/common/base64    [no test files]
?       github.com/hkparker/go-i2p/lib/config   [no test files]
--- FAIL: TestDSA (0.00s)
        dsa_test.go:32: sig="R\x06D\x97\x0fu#\xd1Ҧ\xf3\x11G\x0e\x0e^\xde\ue445\x91\xaf+7\u007f\xbf\xaeC\x1eͩ\x9ctsS\x90\x9b\x0fY\x00"
        dsa_test.go:43: failed: invalid signature
FAIL
coverage: 55.0% of statements
FAIL    github.com/hkparker/go-i2p/lib/crypto   0.065s
?       github.com/hkparker/go-i2p/lib/i2np     [no test files]
?       github.com/hkparker/go-i2p/lib/netdb    [no test files]
?       github.com/hkparker/go-i2p/lib/router   [no test files]
?       github.com/hkparker/go-i2p/lib/transport        [no test files]
?       github.com/hkparker/go-i2p/lib/transport/ssu    [no test files]
?       github.com/hkparker/go-i2p/lib/tunnel   [no test files]
```

--- a/lib/crypto/dsa.go
+++ b/lib/crypto/dsa.go
@@ -162,8 +162,12 @@ func (ds *DSASigner) SignHash(h []byte) (sig []byte, err error) {
 	r, s, err = dsa.Sign(rand.Reader, ds.k, h)
 	if err == nil {
 		sig = make([]byte, 40)
-		copy(sig, r.Bytes())
-		copy(sig[20:], s.Bytes())
+		rb := r.Bytes()
+		rl := len(rb)
+		copy(sig[20-rl:20], rb)
+		sb := s.Bytes()
+		sl := len(sb)
+		copy(sig[20+(20-sl):], sb)
 	}
 	return
 }

--- a/lib/crypto/dsa_test.go
+++ b/lib/crypto/dsa_test.go
@@ -46,10 +46,9 @@ func TestDSA(t *testing.T) {
 	}
 }
 
-
 func BenchmarkDSAGenerate(b *testing.B) {
 	var sk DSAPrivateKey
-	for n := 0 ; n < b.N ; n ++ { 
+	for n := 0; n < b.N; n++ {
 		_, err := sk.Generate()
 		if err != nil {
 			panic(err.Error())
@@ -73,14 +72,14 @@ func BenchmarkDSASignVerify(b *testing.B) {
 	v, _ := pk.NewVerifier()
 	data := make([]byte, 1024)
 	fail := 0
-	for n := 0 ; n < b.N ; n ++ {
+	for n := 0; n < b.N; n++ {
 		sig, err := s.Sign(data)
 		if err != nil {
 			panic(err.Error())
 		}
 		err = v.Verify(data, sig)
 		if err != nil {
-			fail ++
+			fail++
 		}
 	}
 	log.Infof("%d fails %d signs", fail, b.N)

--- a/lib/crypto/dsa_test.go
+++ b/lib/crypto/dsa_test.go
@@ -1,6 +1,7 @@
 package crypto
 
 import (
+	log "github.com/sirupsen/logrus"
 	"testing"
 )
 
@@ -43,4 +44,44 @@ func TestDSA(t *testing.T) {
 		t.Logf("failed: %s", err.Error())
 		t.Fail()
 	}
+}
+
+
+func BenchmarkDSAGenerate(b *testing.B) {
+	var sk DSAPrivateKey
+	for n := 0 ; n < b.N ; n ++ { 
+		_, err := sk.Generate()
+		if err != nil {
+			panic(err.Error())
+		}
+	}
+}
+
+func BenchmarkDSASignVerify(b *testing.B) {
+	var sk DSAPrivateKey
+	var pk DSAPublicKey
+	var err error
+	sk, err = sk.Generate()
+	if err != nil {
+		panic(err.Error())
+	}
+	pk, err = sk.Public()
+	if err != nil {
+		panic(err.Error())
+	}
+	s, _ := sk.NewSigner()
+	v, _ := pk.NewVerifier()
+	data := make([]byte, 1024)
+	fail := 0
+	for n := 0 ; n < b.N ; n ++ {
+		sig, err := s.Sign(data)
+		if err != nil {
+			panic(err.Error())
+		}
+		err = v.Verify(data, sig)
+		if err != nil {
+			fail ++
+		}
+	}
+	log.Infof("%d fails %d signs", fail, b.N)
 }

